Add integration test provider for vmware

jillr opened the pull request:
##### SUMMARY
This change adds a new remote cloud provider option for vmware,
supporting dynamic environments running on worldstream.nl as well as
static environments specific by the user in a cloud-config template.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ansible-test

jillr committed:
Add integration test provider for vmware

This change adds a new remote cloud provider option for vmware,
supporting dynamic environments running on worldstream.nl as well as
static environments specific by the user in a cloud-config template.

jillr committed:
Add integration test provider for vmware

This change adds a new remote cloud provider option for vmware,
supporting dynamic environments running on worldstream.nl as well as
static environments specific by the user in a cloud-config template.

goneri commented on test/integration/cloud-config-vmware.ini.template:
```diff
@@ -0,0 +1,7 @@
+vmware_user: @VMWARE_USER
+vmware_password: @VMWARE_PASSWORD
+vmware_host: @VMWARE_HOST
+vmware_https_port: @VMWARE_HTTPS_PORT
```
The other @foo variables here match the VMWARE_* environment variables. Could we rename this way `VMWARE_PORT` too?

jillr replied:
Yeah, I was mainly looking to differentiate between the https port and the ssh port.  Which tests needed ssh access, I can't find references to how that var is being named today?

goneri commented on test/runner/lib/cloud/vmware.py:
```diff
@@ -0,0 +1,128 @@
+        )
+        ansible_vars.update(dict(parser.items('DEFAULT')))
+
+        return CloudEnvironmentConfig(
```
If we want to stay compatible with the current test-suite, we should expose the following Ansible variables:

- vcenter_hostname: VMWARE_HOST:VMWARE_PORT
- vcenter_username: VMWARE_USER
- vcenter_password: VMWARE_PASSWORD

We can also just exoise the VMWARE_* environment variables for this purpose. I like the idea since this would remove a lot of code. But in this case, we must do some extra adjustments:

- vcenter.py: it should export the same variables
- the test-suite, we should not use vcenter_* variable anymore.

jillr replied:
I don't really see that it removes any code in the provider?  We would just also export the config as env vars at https://github.com/ansible/ansible/blob/0d33aaa88df2bbcc840867b97870d67793e73d76/test/runner/lib/cloud/vmware.py#L118  
It ends up looking like the azure provider, for example.  If env vars are easier for the test targets though that totally works for me and is easy to implement.

goneri replied:
This is the kind of block that won't be necessary anymore:
https://github.com/ansible/ansible/blob/affaedfca8c14f70968a8d2a6c898c488adaa55a/test/integration/targets/vmware_vm_facts/tasks/main.yml#L16-L19

jillr replied:
Ah ok, I didn't realize you meant code in the tests - it is no longer needed with the most recent version then.
https://github.com/jillr/ansible/blob/vmware_provider/test/runner/lib/cloud/vmware.py#L105-L112

jillr committed:
Provide env vars in vmware test provider

jillr committed:
Provide env vars in vmware test provider

mattclay commented on test/runner/lib/cli.py:
```diff
@@ -658,7 +658,7 @@ def add_environments(parser, tox_version=False, tox_only=False):
                         metavar='PROVIDER',
                         help='remote provider to use: %(choices)s',
-                        choices=['default', 'aws', 'azure', 'parallels'],
+                        choices=['default', 'aws', 'azure', 'parallels', 'vmware'],
```
This isn't a general purpose VM provider, so it shouldn't be added here.

mattclay commented on test/integration/cloud-config-vmware.ini.template:
```diff
@@ -0,0 +1,7 @@
+[DEFAULT]
```
We only want a single cloud plugin for vmware/vcenter, so any config should go in the existing `cloud-config-vcenter.ini.template` file instead.

mattclay commented on test/runner/lib/cloud/vmware.py:
```diff
@@ -0,0 +1,120 @@
+"""Vmware plugin for integration tests."""
```
This functionality should go into the existing `vcenter.py` plugin. Use an environment variable to override the current behavior (using the simulator) and use this new logic instead.

mattclay commented on test/runner/lib/cloud/vmware.py:
```diff
@@ -0,0 +1,120 @@
+        """
+        return AnsibleCoreCI(self.args, 'Vmware', 'Vmware',
+                             persist=False, stage=self.args.remote_stage,
+                             provider=self.args.remote_provider)
```
Being able to override the provider here doesn't make any sense, since no other provider supports vmware:

```suggestion
                             provider='vmware')
```

mattclay commented on test/runner/lib/cloud/vmware.py:
```diff
@@ -0,0 +1,120 @@
+        """
+        :rtype: AnsibleCoreCI
+        """
+        return AnsibleCoreCI(self.args, 'Vmware', 'Vmware',
```
I'd stick to lowercase names for consistency with the other implementations:

```suggestion
        return AnsibleCoreCI(self.args, 'vmware', 'vmware',
```

mattclay commented on test/runner/lib/core_ci.py:
```diff
@@ -188,6 +196,25 @@ def _get_parallels_endpoints(self):
 
         raise ApplicationError('Unable to get available endpoints.')
 
+    def _get_vmware_endpoints(self):
```
Supporting multiple endpoints doesn't add value unless thresholds are implemented by the ansible-core-ci service.

jillr replied:
Since we don't have vcenter/clustering deployed/licensed in worldstream atm, we would need to scale by adding additional standalone esxis, thus I figured we'd handle them as individual endpoints.  Is this not the case, or just that you prefer to wait until we actually have multiple hypervisors to worry about it?

mattclay replied:
I figure we wait until we have multiple endpoints before implementing this. Depending on how long it is before that happens, the way we implement this may change.

jillr replied:
Gotcha, thanks for clarifying.

jillr committed:
Update test/runner/lib/cloud/vmware.py

Co-Authored-By: Matt Clay <matt@mystile.com>

jillr committed:
Update test/runner/lib/cloud/vmware.py

Co-Authored-By: Matt Clay <matt@mystile.com>

jillr committed:
Update test/runner/lib/cloud/vmware.py

Co-Authored-By: Matt Clay <matt@mystile.com>

jillr committed:
Update test/runner/lib/cloud/vmware.py

Co-Authored-By: Matt Clay <matt@mystile.com>

jillr committed:
Merge vmware.py w/ vcenter.py test provider

jillr committed:
Merge vmware.py w/ vcenter.py test provider

jillr commented:
@mattclay interested what you think of the general approach for combining the providers that I've started in test/runner/lib/cloud/vcenter.py.

mattclay commented on test/runner/lib/cloud/vcenter.py:
```diff
@@ -12,6 +13,10 @@
 from lib.util import (
     find_executable,
     display,
+    ApplicationError,,
```
```suggestion
    ApplicationError,
```

mattclay commented on test/runner/lib/cloud/vcenter.py:
```diff
@@ -40,6 +53,10 @@ def __init__(self, args):
 
+        # VMWare tests can be run on govcsim or baremetal, either BYO or hosted in worldstream
+        # valid options are None (gocvsim), govcsim, worldstream, static (BYO)
+        self.vmware_test_platform = os.environ['VMWARE_TEST_PLATFORM']
```
```suggestion
        self.vmware_test_platform = os.environ.get('VMWARE_TEST_PLATFORM', '')
```

mattclay commented on test/runner/lib/cloud/vcenter.py:
```diff
@@ -62,10 +79,16 @@ def setup(self):
-        else:
-            self._setup_dynamic()
+        # TODO mostly thinking out loud here, to try out the idea
+        if self.vmware_test_platform is 'govcsim' or None:
```
I'd put this at the end and just make it an else so that it's the default if an invalid (or no option) is chosen.

mattclay commented on test/runner/lib/cloud/vcenter.py:
```diff
@@ -62,10 +79,16 @@ def setup(self):
+        # TODO mostly thinking out loud here, to try out the idea
+        if self.vmware_test_platform is 'govcsim' or None:
+            if self._use_static_config():
+                self._setup_static_simulator()
```
Since there's no static simulator option, it would be convenient to keep the static (bare metal) auto-detection support. If the config file is present, default to static bare metal. If it's not, default to dynamic via govcsim.

The env var would only be needed to force dynamic in the presence of the static config file, or to change the default dynamic provisioning mode.

goneri commented on test/runner/lib/cloud/vcenter.py:
```diff
@@ -139,8 +183,63 @@ def _get_simulator_address(self):
+        self.endpoint = parser.get('DEFAULT', 'vcenter_hostname')
+        self.port = parser.get('DEFAULT', 'vcenter_port')
+
+        if parser.get('DEFAULT', 'vmware_validate_certs') == 'no':
```
I suggest:
 if parser.get('DEFAULT', 'vmware_validate_certs').lower() in ('no', 'false'):

mattclay commented on test/integration/cloud-config-vcenter.ini.template:
```diff
@@ -0,0 +1,7 @@
+[DEFAULT]
```
Include instructions at the top of the file so users know what to do with this. You can use the other templates as starting point.

mattclay commented on test/runner/lib/cloud/vcenter.py:
```diff
@@ -149,13 +248,30 @@ def get_environment_config(self):
-        env_vars = dict(
-            VCENTER_HOST=self._get_cloud_config('vcenter_host'),
-        )
+        vmware_test_platform = os.environ.get('VMWARE_TEST_PLATFORM', '')
```
Environment variables won't be available when delegation such as `--docker` is used. To guarantee settings are available after delegation, use `_set_cloud_config` to preserve them. Example:

https://github.com/ansible/ansible/blob/690037876648e515fdf1e9d6dd82c760b9bd0f72/test/runner/lib/cloud/vcenter.py#L135

To retrieve values use `_get_cloud_config`. Example:

https://github.com/ansible/ansible/blob/690037876648e515fdf1e9d6dd82c760b9bd0f72/test/runner/lib/cloud/vcenter.py#L153

jillr committed:
Improve vcenter test provider

Add explanatory text to cloud-config-vcenter template.
Store and retrieve VMWARE_TEST_PLATFORM envvar with _(set|get)_cloud_config
to ensure it's available when tests have been delegated.

mattclay commented on test/runner/lib/cloud/vcenter.py:
```diff
@@ -40,50 +53,84 @@ def __init__(self, args):
         super(VcenterProvider, self).setup()
 
+        self._set_cloud_config('vmware_test_platform',
+                               os.environ.get('VMWARE_TEST_PLATFORM', ''))
```
You should be able to use `self.vmware_test_platform` here.

mattclay commented on test/runner/lib/cloud/vcenter.py:
```diff
@@ -40,50 +53,84 @@ def __init__(self, args):
 
     def cleanup(self):
         """Clean up the cloud resource and any temporary configuration files after tests complete."""
+        if os.environ.get('VMWARE_TEST_PLATFORM', '') == 'worldstream':
```
You should be able to use `self.vmware_test_platform` here.
